letsencrypt/models/letsencrypt.py

```diff
@@ -2,167 +2,265 @@
-            ]
-            if len(domains) > 1:
-                cmdline.extend([
-                    '-reqexts', 'SAN',
-                ])
-            self.call_cmdline(cmdline)
-        return csr
+                    "or local domains!"))
 
     @api.model
-    def cron(self):
+    def _cron(self):
         domain = urlparse.urlparse(
             self.env['ir.config_parameter'].get_param(
                 'web.base.url', 'localhost')).netloc
-        self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
-                self.env['ir.config_parameter'].get_param(
-                    'letsencrypt.chain_certificate_address',
-                    'https://letsencrypt.org/certs/'
-                    'lets-encrypt-x3-cross-signed.pem')
+        self._validate_domain(domain)
+        # Generate account key
+        account_key_file = self._generate_key('account.key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        client = self._create_client(account_key)
+        new_reg = NewRegistration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        try:
+            client.new_account(new_reg)
+        except errors.ConflictError as e:
+            reg = Registration(key=account_key.public_key())
+            reg_res = RegistrationResource(
+                body=reg,
+                uri=e.location,
             )
-            crt.write(chain_cert.read())
-            chain_cert.close()
-            _logger.info('wrote %s', crt.name)
-        reload_cmd = self.env['ir.config_parameter'].get_param(
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+            client.query_registration(reg_res)
+        csr = self._make_csr(account_key, domain_key_file, domain)
+        authzr = client.new_order(csr)
+        auth_responded = False
+        for authorizations in authzr.authorizations:
+            for challenge in sorted(
+                    authorizations.body.challenges,
+                    key=lambda x: x.chall.typ == TYPE_CHALLENGE_HTTP,
+                    reverse=True):
```

Comment:
What is your point with this? What I am trying to do is have the `TYPE_CHALLENGE_HTTP` come first and the rest of the values do not matter to me.

Reply:
and that's the problem. This means that also in the case of DNS verification, you'll try to do a HTTP verification. *If* there's a DNS provider is configured, the DNS challenge should always win.

Debug through a session where you request a certificate for `therp.eu` and `*.therp.eu`, then you'll see the issue with this: It will needlessly try to do a HTTP verification for `therp.eu`, while it could just do the same DNS verification as for `*.therp.eu`. This btw should be your test case and we only need to talk again when you've successfully requested a certificate for that on our helpdesk.

Reply:
> This means that also in the case of DNS verification, you'll try to do a HTTP verification. If there's a DNS provider is configured, the DNS challenge should always win.

Letsencrypt always returns the challenges that are the '**least common denominator**' for all the domains in the CSR.
If `therp.eu` and `*.therp.eu` exist on the same CSR, **only a DNS challenge will be returned**

```
letsencrypt: Making CSR for the following domains: [u'therp.eu', u'*.therp.eu', u'helptes
t10.therp.nl']                                                                           
> /home/george/therp/projects/customer/oe_bvtherp/parts/server-tools/letsencrypt/models/l
etsencrypt.py(124)_cron()                                                                
-> for challenge in sorted(                                                              
(Pdb) authorizations.body.challenges                                                     
(ChallengeBody(chall=DNS01(token='?\xea\xe2\x8b\xb1s\xb5\xc5\x92+\xa7\xd2\x8bMW\xcc\xd9\x
b8\xadf\x16\xd1\xfeV\xbf\x84\xd2&\xd9\xa1\x95\xe8'), status=Status(pending), uri=u'https:
//acme-staging-v02.api.letsencrypt.org/acme/challenge/RYDHjwhNh4360S2C7vs5hrW9Bj0XR6lKTn4
RwAKY3q8/302776337', validated=None, _url=u'https://acme-staging-v02.api.letsencrypt.org/
acme/challenge/RYDHjwhNh4360S2C7vs5hrW9Bj0XR6lKTn4RwAKY3q8/302776337', error=None),)     
(Pdb) len(authorizations.body.challenges)                                                
1                                                                                        
```

I prefer the "most easy to satisfy" challenge. If letsencrypt says that an http challenge is all it needs to validate *all* the domains we have then so be it. Else just move on to dns.

Reply:
[citation needed]

my tests had different results, there it returned all three challenges for therp.eu. And HTTP is by no means easier to satisfy, think for example of an installation that's not accessible from the internet at all or only via VPN etc.

Reply:
I had written a big reply here but a crash on my wayland made me lose it. Anyway this is how it went:

1) 

> Letsencrypt always returns the challenges that are the 'least common denominator' for all the domains in the CSR.
> If therp.eu and *.therp.eu exist on the same CSR, only a DNS challenge will be returned

I retract this since it is downright wrong. 

A domain is an identifier and the server returns an authorization per identifier. And that authorization has it's challenges. `therp.eu` and `*.therp.eu` are different and at least one of the challenges must be satisfied. If I use DNS on both then I will have two TXT records. Also, the DNS changes take a while.

Also, when it comes to polling, the server MUST return Retry-After so that we can use that to retry. As seen here https://github.com/letsencrypt/boulder/blob/c55cabfc3eb453d80745b81f6a1ff20cd1fd3b8b/docs/acme-divergences.md#section-82
this is not yet implemented (no references by greping on master as well)

The acme library does not implement this as well and they put it on a wishlist here https://github.com/certbot/certbot/issues/7005. I was ready to create a PR for this, but since Boulder does not implement it it would not make sense for us. (Got a book on golang, that seems like a great opportunity get a grip on this language)

Nonetheless, I did implement the parameter you requested (in a slightly different way), but I am having difficulties with the `therp.eu` domain. It should redirect to vanaheim (as I have set the ip on the A record but that does not seem to be the case.)

Reply:
concerning the parameter: Check out my comment in https://github.com/OCA/server-tools/pull/1194/commits/231c0fb34db469eeb89b158b02a62087dec5ebe9#r251449875 - I think you really need a loop here where you to a `sleep` for this amount of time, certbot does the same thing.

Concerning the therp.eu domain: If I read https://github.com/OCA/l10n-netherlands/pull/179/files#diff-608ca28d494b87177e992bf1228a4513R43 and https://github.com/benkonrath/transip-api/blob/master/transip/service/domain.py#L147 correctly, you wipe all domain entries in this call and replace them with your new entry. Seems you need to use the API for querying all entries, and appending yours to this list. And please restore the previously existing therp.eu subdomains.

Reply:
> Seems you need to use the API for querying all entries, and appending yours to this list

A PR has been made against the repository of the library owner here when this is merged I will update the pin and make the change here:
https://github.com/benkonrath/transip-api/pull/48

> concerning the parameter: Check out my comment in 231c0fb#r251449875 - I think you really need a loop here where you to a sleep for this amount of time, certbot does the same thing.

This is what the library is doing when I pass the `deadline` here, it uses sleep and loops https://github.com/certbot/certbot/blob/0.33.x/acme/acme/client.py#L723 why should we do it in another way and outside of the library? Also, the link you posted does not contain any comment.

Reply:
loop: Last time I tested this, letsencrypt didn't return a pending state but an invalid state. If this is fixed by now, then that's fine. But test this. My comment is on https://github.com/OCA/server-tools/pull/1194/files#r251449875 - if the mail shows the wrong link, open it from the github interface. This also contains a link to certbot who also works around this issue.

For your patch: I'd suggest to add some text in the PR

Reply:
Here is the issue with this:

If you try to satisfy a dns request and for some reason fails, all the subsequent dns challenges will immediately fail. **This happens because boulder "sees" the old TXT record and does not wait at all, rather it fails immediately.** If it were no TXT record it would wait for it to appear. If that does not happen, the certificate should work. I think this is an issue of their server, the challenge should not be invalided that quickly and should wait in any case (there is no way for an instant DNS update), also multiple TXT records with the same name are allow which can cause further issues. I am studying golang and will make contributions to boulder at some point.

Since you were not able to see the reason your validations failed I made another PR to acme this time
https://github.com/certbot/certbot/pull/7066

Reply:
but still we need to work around the issue for now, right?